// @ngx-formly/core/esm2015/lib/templates/field.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Input } from '@angular/core';
import { FormGroup } from '@angular/forms';
/**
 * @abstract
 */
export class Field {
    /**
     * @return {?}
     */
    get key() { return this.field.key; }
    /**
     * @return {?}
     */
    get formControl() { return this.field.formControl; }
    /**
     * @return {?}
     */
    get to() { return this.field.templateOptions; }
    /**
     * @return {?}
     */
    get showError() { return this.options.showError(this); }
    /**
     * @return {?}
     */
    get id() { return this.field.id; }
    /**
     * @return {?}
     */
    get formState() { return this.options.formState || {}; }
}
Field.propDecorators = {
    form: [{ type: Input }],
    field: [{ type: Input }],
    model: [{ type: Input }],
    options: [{ type: Input }]
};
function Field_tsickle_Closure_declarations() {
    /** @type {?} */
    Field.prototype.form;
    /** @type {?} */
    Field.prototype.field;
    /** @type {?} */
    Field.prototype.model;
    /** @type {?} */
    Field.prototype.options;
}
