// @ngx-formly/bootstrap/esm2015/lib/wrappers/message-validation.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ViewChild, ViewContainerRef } from '@angular/core';
import { FieldWrapper } from '@ngx-formly/core';
export class FormlyWrapperValidationMessages extends FieldWrapper {
    /**
     * @return {?}
     */
    get validationId() {
        return this.field.id + '-message';
    }
}
FormlyWrapperValidationMessages.decorators = [
    { type: Component, args: [{
                selector: 'formly-wrapper-validation-messages',
                template: `
    <ng-template #fieldComponent></ng-template>
    <div *ngIf="showError">
      <small class="text-danger invalid-feedback" [style.display]="'block'" role="alert" [id]="validationId">
        <formly-validation-message [field]="field"></formly-validation-message>
      </small>
    </div>
  `,
            },] },
];
FormlyWrapperValidationMessages.propDecorators = {
    fieldComponent: [{ type: ViewChild, args: ['fieldComponent', { read: ViewContainerRef },] }]
};
function FormlyWrapperValidationMessages_tsickle_Closure_declarations() {
    /** @type {?} */
    FormlyWrapperValidationMessages.prototype.fieldComponent;
}
