// @ngx-formly/core/esm5/lib/services/formly.form.expression.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { FormGroup, FormArray } from '@angular/forms';
import { evalExpression, FORMLY_VALIDATORS, getFieldModel, isObject, getKeyPath, isNullOrUndefined } from '../utils';
/**
 * \@internal
 */
var FormlyFormExpression = /** @class */ (function () {
    function FormlyFormExpression() {
    }
    /**
     * @param {?} form
     * @param {?=} fields
     * @param {?=} model
     * @param {?=} options
     * @return {?}
     */
    FormlyFormExpression.prototype.checkFields = /**
     * @param {?} form
     * @param {?=} fields
     * @param {?=} model
     * @param {?=} options
     * @return {?}
     */
    function (form, fields, model, options) {
        if (fields === void 0) { fields = []; }
        this._checkFields(form, fields, model, options);
    };
    /**
     * @param {?} form
     * @param {?=} fields
     * @param {?=} model
     * @param {?=} options
     * @return {?}
     */
    FormlyFormExpression.prototype._checkFields = /**
     * @param {?} form
     * @param {?=} fields
     * @param {?=} model
     * @param {?=} options
     * @return {?}
     */
    function (form, fields, model, options) {
        var _this = this;
        if (fields === void 0) { fields = []; }
        fields.forEach(function (field) {
            _this.checkFieldExpressionChange(form, field, _this.getParentModel(model, field), options);
            _this.checkFieldVisibilityChange(form, field, _this.getParentModel(model, field), options);
            if (field.fieldGroup && field.fieldGroup.length > 0) {
                _this._checkFields(field.formControl ? /** @type {?} */ (field.formControl) : form, field.fieldGroup, _this.getParentModel(model, field), options);
            }
        });
    };
    /**
     * @param {?} form
     * @param {?} field
     * @param {?} model
     * @param {?} options
     * @return {?}
     */
    FormlyFormExpression.prototype.checkFieldExpressionChange = /**
     * @param {?} form
     * @param {?} field
     * @param {?} model
     * @param {?} options
     * @return {?}
     */
    function (form, field, model, options) {
        if (!field || !field.expressionProperties) {
            return;
        }
        var /** @type {?} */ expressionProperties = field.expressionProperties;
        var /** @type {?} */ validators = FORMLY_VALIDATORS.map(function (v) { return "templateOptions." + v; });
        for (var /** @type {?} */ key in expressionProperties) {
            var /** @type {?} */ expressionValue = evalExpression(expressionProperties[key].expression, { field: field }, [model, options.formState]);
            if (expressionProperties[key].expressionValue !== expressionValue
                && (!isObject(expressionValue) || JSON.stringify(expressionValue) !== JSON.stringify(expressionProperties[key].expressionValue))) {
                expressionProperties[key].expressionValue = expressionValue;
                evalExpression(expressionProperties[key].expressionValueSetter, { field: field }, [expressionValue, model, field]);
                if (key.indexOf('model.') === 0) {
                    var /** @type {?} */ path = key.replace(/^model\./, ''), /** @type {?} */
                    control = field.key && key === path ? field.formControl : form.get(path);
                    if (control
                        && !(isNullOrUndefined(control.value) && isNullOrUndefined(expressionValue))
                        && control.value !== expressionValue) {
                        control.patchValue(expressionValue);
                    }
                }
                if (validators.indexOf(key) !== -1 && field.formControl) {
                    field.formControl.updateValueAndValidity({ emitEvent: false });
                }
            }
        }
    };
    /**
     * @param {?} form
     * @param {?} field
     * @param {?} model
     * @param {?} options
     * @return {?}
     */
    FormlyFormExpression.prototype.checkFieldVisibilityChange = /**
     * @param {?} form
     * @param {?} field
     * @param {?} model
     * @param {?} options
     * @return {?}
     */
    function (form, field, model, options) {
        if (!field || isNullOrUndefined(field.hideExpression)) {
            return;
        }
        var /** @type {?} */ hideExpressionResult = !!evalExpression(field.hideExpression, { field: field }, [model, options.formState]);
        if (hideExpressionResult !== field.hide) {
            // toggle hide
            field.hide = hideExpressionResult;
            field.templateOptions.hidden = hideExpressionResult;
            if (field.formControl && field.key) {
                var /** @type {?} */ parent_1 = this.fieldParentFormControl(form, field);
                if (parent_1) {
                    var /** @type {?} */ control = parent_1.get("" + this.fieldKey(field));
                    if (hideExpressionResult === true && control) {
                        this.removeFieldControl(parent_1, field);
                    }
                    else if (hideExpressionResult === false && !control) {
                        this.addFieldControl(parent_1, field, model);
                    }
                }
            }
            if (options.fieldChanges) {
                options.fieldChanges.next(/** @type {?} */ ({ field: field, type: 'hidden', value: hideExpressionResult }));
            }
        }
    };
    /**
     * @param {?} parent
     * @param {?} field
     * @param {?} model
     * @return {?}
     */
    FormlyFormExpression.prototype.addFieldControl = /**
     * @param {?} parent
     * @param {?} field
     * @param {?} model
     * @return {?}
     */
    function (parent, field, model) {
        var /** @type {?} */ fieldModel = this.getFieldModel(model, field);
        if (!(isNullOrUndefined(field.formControl.value) && isNullOrUndefined(fieldModel))
            && field.formControl.value !== fieldModel) {
            field.formControl.patchValue(fieldModel, { emitEvent: false });
        }
        if (parent instanceof FormArray) {
            parent.push(field.formControl);
        }
        else if (parent instanceof FormGroup) {
            parent.addControl("" + this.fieldKey(field), field.formControl);
        }
    };
    /**
     * @param {?} model
     * @param {?} field
     * @return {?}
     */
    FormlyFormExpression.prototype.getFieldModel = /**
     * @param {?} model
     * @param {?} field
     * @return {?}
     */
    function (model, field) {
        if (field.fieldGroup || field.fieldArray) {
            return model;
        }
        return getFieldModel(model, field, false);
    };
    /**
     * @param {?} model
     * @param {?} field
     * @return {?}
     */
    FormlyFormExpression.prototype.getParentModel = /**
     * @param {?} model
     * @param {?} field
     * @return {?}
     */
    function (model, field) {
        if (field.key && (field.fieldGroup || field.fieldArray)) {
            return getFieldModel(model, field, true);
        }
        return model;
    };
    /**
     * @param {?} parent
     * @param {?} field
     * @return {?}
     */
    FormlyFormExpression.prototype.removeFieldControl = /**
     * @param {?} parent
     * @param {?} field
     * @return {?}
     */
    function (parent, field) {
        if (parent instanceof FormArray) {
            parent.removeAt(/** @type {?} */ (this.fieldKey(field)));
        }
        else if (parent instanceof FormGroup) {
            parent.removeControl("" + this.fieldKey(field));
        }
    };
    /**
     * @param {?} form
     * @param {?} field
     * @return {?}
     */
    FormlyFormExpression.prototype.fieldParentFormControl = /**
     * @param {?} form
     * @param {?} field
     * @return {?}
     */
    function (form, field) {
        var /** @type {?} */ paths = getKeyPath(field);
        paths.pop(); // remove last path
        return /** @type {?} */ ((paths.length > 0 ? form.get(paths) : form));
    };
    /**
     * @param {?} field
     * @return {?}
     */
    FormlyFormExpression.prototype.fieldKey = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        return getKeyPath(field).pop();
    };
    FormlyFormExpression.decorators = [
        { type: Injectable },
    ];
    return FormlyFormExpression;
}());
export { FormlyFormExpression };
