// @ngx-formly/core/esm5/lib/templates/field.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Input } from '@angular/core';
import { FormGroup } from '@angular/forms';
/**
 * @abstract
 */
var Field = /** @class */ (function () {
    function Field() {
    }
    Object.defineProperty(Field.prototype, "key", {
        get: /**
         * @return {?}
         */
        function () { return this.field.key; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Field.prototype, "formControl", {
        get: /**
         * @return {?}
         */
        function () { return this.field.formControl; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Field.prototype, "to", {
        get: /**
         * @return {?}
         */
        function () { return this.field.templateOptions; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Field.prototype, "showError", {
        get: /**
         * @return {?}
         */
        function () { return this.options.showError(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Field.prototype, "id", {
        get: /**
         * @return {?}
         */
        function () { return this.field.id; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Field.prototype, "formState", {
        get: /**
         * @return {?}
         */
        function () { return this.options.formState || {}; },
        enumerable: true,
        configurable: true
    });
    Field.propDecorators = {
        form: [{ type: Input }],
        field: [{ type: Input }],
        model: [{ type: Input }],
        options: [{ type: Input }]
    };
    return Field;
}());
export { Field };
function Field_tsickle_Closure_declarations() {
    /** @type {?} */
    Field.prototype.form;
    /** @type {?} */
    Field.prototype.field;
    /** @type {?} */
    Field.prototype.model;
    /** @type {?} */
    Field.prototype.options;
}
